recipe/2024-02-terraform-providers.md

```diff
@@ -0,0 +1,190 @@
+2. Enable users to use terraform recipes with multiple configurations for the same provider. 
+3. Enable provider configurations to be available for all recipies used in an environment.
+
+
+### Non goals
+1. Updates to Bicep Provider configuration.
+2. Authentication for providers hosted in private registries.
+3. Support for .terraformrc files.
+
+### User scenarios (optional)
+
+#### User story 1
+As an operator, I maintain a set of Terraform recipes for use with Radius. I have a set of provider configurations which are applicable across multiple recipes and I would like to configure them in a single centralized location for ease of maintenance.
+
+
+#### User story 2
+As an operator, I would like to manage cloud resources in different geographical regions. To enable this, I need to set up multiple configurations for the same provider for different regions and use an alias to refer to individual provider configurations. 
+
+## Design
+### Design details
+
+Reviewing popular provider configurations (GCP, Oracle, Heroku, DigitalOcean, Docker etc.) a large number of provider configurations can be setup by handling a combination of key value pairs, key value pairs in nested objects, secrets and environment variables.
+
+### Key-value pairs
+Key-value pairs will be parsed and saved to a Terraform configuration file.
+
+### Secrets
+Secrets will be handled similarly to the approach described in document [Design document for Private Terraform Repository](https://github.com/radius-project/design-notes/blob/3644b754152edc97e641f537b20cf3d87a386c43/recipe/2024-01-support-private-terraform-repository.md) wherein Applications.Core/secretStores can point to an existing K8s secret.
+
+The system will call the ListSecrets() api in Applications.Core namespace, retrieve contents of the secret and build the Terraform provider configuration.
+
+### Environment variables
+In a significant number of providers, as per documentation, environment variables are used one of the methods of saving sensitive credential data along with insensitive data. We allow the users to set environment variables for provider configuration. For sensitive information, we recommend the users save these values as secrets and point to them inside the env block.
+
+```
+...
+ recipeConfig: {
+    terraform: {
+      ...
+     providers: [...]
+     env: {
+       key: value
+       secrets: {
+        source: secretStoreOCICred.id
+      }
+     }
+```
+
+Environment variables apply to all providers configured in the environment. The system cannot set two separate values for the same environment variable for multple provider configurations. In such cases, per provider documentation, there may be alternatives that users can avail (eg. For GCP, users can set  *credentials* field inside each instance of provider config as opposed to using env variable GOOGLE_APPLICATION_CREDENTIALS).
```

Comment:
Would we recommend that users avoid using env-vars in favor of other techniques?

Reply:
Use of env-vars would be the users' decision i think. We're outlining details on what the system can support. I'm not sure we want to have a blanket recommendation in this document. Each provider may have their recommended way of setting up configuration.